src/decode-arp.c

```diff
@@ -0,0 +1,72 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ *
+ * \author Giuseppe Longo <giuseppe@glongo.it>
+ *
+ */
+
+#include "suricata-common.h"
+#include "decode.h"
+#include "decode-arp.h"
+#include "decode-events.h"
+
+int DecodeARP(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p, const uint8_t *pkt, uint32_t len)
+{
+    StatsIncr(tv, dtv->counter_arp);
+
+    if (unlikely(len < ARP_HEADER_MIN_LEN)) {
+        ENGINE_SET_INVALID_EVENT(p, ARP_PKT_TOO_SMALL);
+        return TM_ECODE_FAILED;
+    }
+
+    if (!PacketIncreaseCheckLayers(p)) {
+        return TM_ECODE_FAILED;
+    }
+
+    const ARPHdr *arph = PacketSetARP(p, pkt);
+    if (unlikely(arph == NULL))
+        return TM_ECODE_FAILED;
+
+    if (SCNtohs(arph->hw_type) != ARP_HW_TYPE_ETHERNET) {
+        ENGINE_SET_INVALID_EVENT(p, ARP_UNSUPPORTED_HARDWARE);
+        return TM_ECODE_FAILED;
+    }
+
+    if (SCNtohs(arph->proto_type) != ETHERNET_TYPE_IP) {
+        ENGINE_SET_INVALID_EVENT(p, ARP_UNSUPPORTED_PROTOCOL);
+        return TM_ECODE_FAILED;
+    }
+
+    if (unlikely(len != ARP_HEADER_LEN)) {
+        ENGINE_SET_INVALID_EVENT(p, ARP_UNSUPPORTED_PKT);
```

Comment:
if we get here, we know that packet is arp, hw type ethernet, proto type ip. So is a size mismatch here a "invalid" or an "unsupported" packet? I think invalid seems more appropriate?

Reply:
in fact, I would think < ARP_HEADER_LEN => invalid
and > ARP_HEADER_LEN "trailing data" or something?

In general, it might be good to validate the fixed size part of the header, so the first 8 bytes
```c
    uint16_t hw_type;
    uint16_t proto_type;
    uint8_t hw_size;
    uint8_t proto_size;
    uint16_t opcode;
```
`+` the sum of the hw_size and proto_size, and see if it matches the input. Shorter is bad/invalid+event, longer is suspicious/valid+event?

Reply:
> if we get here, we know that packet is arp, hw type ethernet, proto type ip. So is a size mismatch here a "invalid" or an "unsupported" packet? I think invalid seems more appropriate?

Yes, I think invalid is more appropriate.

Reply:
> in fact, I would think < ARP_HEADER_LEN => invalid and > ARP_HEADER_LEN "trailing data" or something?
> 
> In general, it might be good to validate the fixed size part of the header, so the first 8 bytes
> 
> ```c
>     uint16_t hw_type;
>     uint16_t proto_type;
>     uint8_t hw_size;
>     uint8_t proto_size;
>     uint16_t opcode;
> ```
> 
> `+` the sum of the hw_size and proto_size, and see if it matches the input. Shorter is bad/invalid+event, longer is suspicious/valid+event?

I've found a pcap where after the ARP pkt, 18 bytes are added as padding for Ethernet (which can be filtered with eth.padding). Here is how the packet looks:

```
ARP:
00 01 08 00 06 04 00 03  00 00 a1 12 dd 88 00 00
00 00 00 00 a1 12 dd 88  00 00 00 00
Ethernet padding:
00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
00 00
```

To summarize, if the length is shorter than ARP_HEADER_LEN, it is invalid (and an event is set). 
If it is longer, it is valid, but I wonder if we should set an event, as it can be more noisy than helpful if padding is often added at the end of the packet.